File: scripts/test-acceptance.ts
import {
  validateSecretStrength,
  createSessionToken,
  verifySessionToken,
  createServiceJwt,
  verifyServiceJwt,
  getAuthSession,
  DISALLOWED_DEFAULT_SECRETS
} from '../src/lib/auth';
import { db } from '../src/lib/db';
import { seedDatabaseIfEmpty } from '../src/lib/db/seed';
import { isPrivateIp, validateSafeUrl, safeFetch } from '../src/lib/utils/safe-fetch';
import { commerceEngine } from '../src/lib/commerce';
import { executeTool } from '../src/lib/tools';
import { SignJWT } from 'jose';

async function main() {
  console.log('================================================================');
  console.log('  RUNNING PRODUCTION ACCEPTANCE TEST SUITE (8/8 CRITERIA)');
  console.log('================================================================\n');

  let passed = 0;
  let total = 0;

  function assert(condition: boolean, testName: string, detail?: string) {
    total++;
    if (condition) {
      console.log(`[PASS] Test ${total}: ${testName}`);
      passed++;
    } else {
      console.error(`[FAIL] Test ${total}: ${testName}${detail ? ` - ${detail}` : ''}`);
      throw new Error(`Acceptance Test Failed: ${testName}`);
    }
  }

  // Ensure test environment variables with strong secrets
  process.env.SESSION_JWT_SECRET = 'super_secure_production_session_jwt_secret_987654321_aaas';
  process.env.SERVICE_JWT_SECRET = 'super_secure_production_service_jwt_secret_123456789_aaas';
  process.env.INTERNAL_SERVICE_SECRET = process.env.SERVICE_JWT_SECRET;

  // --------------------------------------------------------------------------
  // CRITERION 1: Secret & Token Validation
  // --------------------------------------------------------------------------
  console.log('\n--- Criterion 1: Secrets & Token Validation ---');

  // Test weak secrets rejection
  let caughtWeak = false;
  try {
    validateSecretStrength('short_secret', 'TEST_SECRET');
  } catch {
    caughtWeak = true;
  }
  assert(caughtWeak, 'Rejects secrets shorter than 32 characters');

  let caughtDisallowed = false;
  try {
    validateSecretStrength(DISALLOWED_DEFAULT_SECRETS[0], 'TEST_SECRET');
  } catch {
    caughtDisallowed = true;
  }
  assert(caughtDisallowed, 'Rejects known insecure repository default secret');

  // Test token signed with old default string is rejected
  const oldSecret = new TextEncoder().encode('super_secret_jwt_key_enterprise_grade_aaas_platform_2026');
  const oldForgedToken = await new SignJWT({ userId: 'usr_merchant_01', email: 'merchant@shopmate.com' })
    .setProtectedHeader({ alg: 'HS256' })
    .setIssuedAt()
    .setExpirationTime('1h')
    .sign(oldSecret);

  const verifiedOld = await verifySessionToken(oldForgedToken);
  assert(verifiedOld === null, 'Token signed with old default secret string is rejected (returns null / 401)');

  // Test valid session token issuance and verification with distinct iss/aud
  const validSessionToken = await createSessionToken({
    userId: 'usr_test_01',
    email: 'tester@acme.com',
    workspaceId: 'ws_tenant_a'
  });
  const verifiedSession = await verifySessionToken(validSessionToken);
  assert(
    verifiedSession !== null && verifiedSession.userId === 'usr_test_01' && verifiedSession.workspaceId === 'ws_tenant_a',
    'Valid session token with iss:aaas-auth and aud:aaas-app verifies successfully'
  );

  // Test service token issuance and verification
  const validServiceToken = await createServiceJwt('ws_tenant_a', 'service_node_01', 'ADMIN');
  const verifiedService = await verifyServiceJwt(validServiceToken);
  assert(
    verifiedService !== null && verifiedService.workspace_id === 'ws_tenant_a',
    'Service JWT with iss:aaas-node and aud:aaas-python verifies successfully'
  );

  // --------------------------------------------------------------------------
  // CRITERION 2: Production Demo Account Suppression
  // --------------------------------------------------------------------------
  console.log('\n--- Criterion 2: Production Demo Account Isolation ---');
  // Clear DB memory to simulate fresh production environment
  db.users.length = 0;
  db.workspaces.length = 0;
  db.workspace_members.length = 0;

  const prodUser = db.users.find(u => u.email === 'admin@aaas-platform.com');
  assert(prodUser === undefined, 'In unseeded database, demo account admin@aaas-platform.com does not exist');

  // Explicit seeding runs only when requested
  await seedDatabaseIfEmpty(true);
  const seededMerchant = db.users.find(u => u.email === 'merchant@shopmate.com');
  assert(seededMerchant !== undefined, 'Explicit seed populates demo accounts on command');

  // --------------------------------------------------------------------------
  // CRITERION 3: Anonymous Requests Never Authenticate
  // --------------------------------------------------------------------------
  console.log('\n--- Criterion 3: Anonymous Requests Denial ---');
  const anonSession = await getAuthSession();
  assert(anonSession === null, 'Anonymous request returns null session (401), zero first-user dev fallback');

  const emptyReq = new Request('http://localhost:3000/api/dashboard', {
    headers: {}
  });
  const reqSession = await getAuthSession(emptyReq);
  assert(reqSession === null, 'Request with no Authorization header or session cookie returns null session');

  // --------------------------------------------------------------------------
  // CRITERION 4: Workspace Membership Isolation
  // --------------------------------------------------------------------------
  console.log('\n--- Criterion 4: Workspace Membership Isolation ---');
  // Create Tenant A and Tenant B
  const userTenantA = {
    id: 'usr_tenant_a',
    email: 'user_a@tenant-a.com',
    name: 'User A',
    password_hash: 'hash',
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString()
  };
  db.users.push(userTenantA);

  const wsTenantA = {
    id: 'ws_tenant_a',
    name: 'Tenant A Workspace',
    slug: 'tenant-a',
    plan: 'PRO' as const,
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString(),
    settings: { retention_days: 30 }
  };
  const wsTenantB = {
    id: 'ws_tenant_b',
    name: 'Tenant B Workspace',
    slug: 'tenant-b',
    plan: 'PRO' as const,
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString(),
    settings: { retention_days: 30 }
  };
  db.workspaces.push(wsTenantA, wsTenantB);

  // User A is ONLY member of Tenant A
  db.workspace_members.push({
    id: 'wsm_a_01',
    workspace_id: 'ws_tenant_a',
    user_id: 'usr_tenant_a',
    role: 'ADMIN',
    created_at: new Date().toISOString()
  });

  // Token requesting Tenant B
  const crossTenantToken = await createSessionToken({
    userId: 'usr_tenant_a',
    email: 'user_a@tenant-a.com',
    workspaceId: 'ws_tenant_b'
  });

  const crossReq = new Request('http://localhost:3000/api/data', {
    headers: { Authorization: `Bearer ${crossTenantToken}` }
  });
  const crossSession = await getAuthSession(crossReq);
  assert(crossSession === null, 'Non-member accessing workspace B is denied (null/403), zero fallback to workspaces[0]');

  // Token requesting Tenant A (authorized)
  const legitToken = await createSessionToken({
    userId: 'usr_tenant_a',
    email: 'user_a@tenant-a.com',
    workspaceId: 'ws_tenant_a'
  });
  const legitReq = new Request('http://localhost:3000/api/data', {
    headers: { Authorization: `Bearer ${legitToken}` }
  });
  const legitSession = await getAuthSession(legitReq);
  assert(legitSession !== null && legitSession.workspaceId === 'ws_tenant_a', 'Authorized member accessing workspace A succeeds');

  // --------------------------------------------------------------------------
  // CRITERION 5: SSRF & Ingestion Hardening
  // --------------------------------------------------------------------------
  console.log('\n--- Criterion 5: SSRF Protection ---');
  assert(isPrivateIp('169.254.169.254'), 'Blocks AWS/GCP cloud metadata IP 169.254.169.254');
  assert(isPrivateIp('172.16.0.5'), 'Blocks RFC 1918 private IP 172.16.0.5');
  assert(isPrivateIp('::1'), 'Blocks IPv6 loopback ::1');
  assert(isPrivateIp('[::1]'), 'Blocks bracketed IPv6 loopback [::1]');
  assert(isPrivateIp('0x7f000001'), 'Blocks hex representation of loopback 0x7f000001');
  assert(isPrivateIp('2130706433'), 'Blocks integer dword representation of loopback 2130706433');
  assert(isPrivateIp('::ffff:127.0.0.1'), 'Blocks IPv4-mapped IPv6 loopback ::ffff:127.0.0.1');
  assert(isPrivateIp('::ffff:169.254.169.254'), 'Blocks IPv4-mapped IPv6 metadata ::ffff:169.254.169.254');

  let blockedFileProto = false;
  try {
    await validateSafeUrl('file:///etc/passwd');
  } catch {
    blockedFileProto = true;
  }
  assert(blockedFileProto, 'Rejects non-HTTP protocols (file:///etc/passwd)');

  let blockedMetadata = false;
  try {
    await validateSafeUrl('http://169.254.169.254/latest/meta-data/');
  } catch {
    blockedMetadata = true;
  }
  assert(blockedMetadata, 'Blocks direct request to metadata endpoint');

  // --------------------------------------------------------------------------
  // CRITERION 6: Zero Fabricated Policy Text
  // --------------------------------------------------------------------------
  console.log('\n--- Criterion 6: Zero Fabricated Knowledge / Empty Tenant Safety ---');
  // Empty workspace has 0 chunks
  const emptyTenantChunks = db.knowledge_chunks.filter(c => c.workspace_id === 'ws_tenant_b');
  assert(emptyTenantChunks.length === 0, 'Tenant B has 0 knowledge chunks');

  // --------------------------------------------------------------------------
  // CRITERION 7: Order Lookup with Customer Email Matching
  // --------------------------------------------------------------------------
  console.log('\n--- Criterion 7: Order Lookup with Customer Email Matching ---');
  // Test Sarah Connor's order #10482
  const orderCorrect = await commerceEngine.getOrder('ws_acme_corp', '#10482', 'sarah.connor@example.com');
  assert(orderCorrect !== null && orderCorrect.customer_email === 'sarah.connor@example.com', 'Order lookup with correct matching email succeeds');

  const orderWrongEmail = await commerceEngine.getOrder('ws_acme_corp', '#10482', 'attacker@evil.com');
  assert(orderWrongEmail === null, 'Order lookup with mismatched email returns null (identical 404, prevents enumeration)');

  const orderMissingEmail = await commerceEngine.getOrder('ws_acme_corp', '#10482', '');
  assert(orderMissingEmail === null, 'Order lookup with missing email returns null');

  // Test tool execution requiring customer_email
  const toolResultWrong = await executeTool({
    tool_id: 'order_lookup',
    workspace_id: 'ws_acme_corp',
    agent_id: 'agent_shopmate_01',
    conversation_id: 'conv_1',
    parameters: { order_number: '#10482', customer_email: 'wrong@mail.com' }
  });
  assert(toolResultWrong.status === 'FAILED', 'Tool execution with mismatched email fails with order not found message');

  const toolResultCorrect = await executeTool({
    tool_id: 'order_lookup',
    workspace_id: 'ws_acme_corp',
    agent_id: 'agent_shopmate_01',
    conversation_id: 'conv_1',
    parameters: { order_number: '#10482', customer_email: 'sarah.connor@example.com' }
  });
  assert(toolResultCorrect.status === 'SUCCESS' && toolResultCorrect.data.order_number === '#10482', 'Tool execution with correct email succeeds');

  console.log('\n================================================================');
  console.log(`  ACCEPTANCE TEST SUMMARY: ${passed}/${total} TESTS PASSED (100%)`);
  console.log('================================================================\n');
}

main().catch((err) => {
  console.error('Test suite failed:', err);
  process.exit(1);
});

File: python-backend/test_acceptance.py
import os
import sys
import jwt
import asyncio
import unittest

# Set valid production-grade secrets for test execution
VALID_SERVICE_SECRET = "super_secure_production_service_jwt_secret_123456789_aaas"
os.environ["INTERNAL_SERVICE_SECRET"] = VALID_SERVICE_SECRET
os.environ["SERVICE_JWT_SECRET"] = VALID_SERVICE_SECRET
os.environ["DATABASE_URL"] = "sqlite+aiosqlite:///./test_acceptance.db"

from app.auth import decode_token, verify_service_jwt, get_service_secret, DISALLOWED_DEFAULT_SECRETS
from app.tools import lookup_order, _fetch_order_db
from app.rag import execute_rag_pipeline
from app.db.database import init_db
from fastapi import HTTPException

class TestAcceptanceHardening(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        asyncio.run(init_db())

    def test_01_service_secret_strength_validation(self):
        """Service secret rejects missing, short, or known insecure strings."""
        old_env = os.environ.get("INTERNAL_SERVICE_SECRET")
        try:
            # Short secret
            os.environ["INTERNAL_SERVICE_SECRET"] = "too_short"
            with self.assertRaises(RuntimeError):
                get_service_secret()

            # Disallowed repo default secret
            os.environ["INTERNAL_SERVICE_SECRET"] = DISALLOWED_DEFAULT_SECRETS[0]
            with self.assertRaises(RuntimeError):
                get_service_secret()
        finally:
            os.environ["INTERNAL_SERVICE_SECRET"] = old_env

    def test_02_old_default_token_rejected_with_401(self):
        """Tokens signed with old default secret string are rejected with 401."""
        old_secret = "super_secret_jwt_key_enterprise_grade_aaas_platform_2026"
        old_token = jwt.encode(
            {"workspace_id": "ws_acme_corp", "sub": "attacker", "exp": 9999999999},
            old_secret,
            algorithm="HS256"
        )
        with self.assertRaises(HTTPException) as ctx:
            decode_token(old_token)
        self.assertEqual(ctx.exception.status_code, 401)

    def test_03_valid_service_jwt_verification(self):
        """Valid service JWT with iss:aaas-node, aud:aaas-python, exp succeeds."""
        import time
        token = jwt.encode(
            {
                "workspace_id": "ws_acme_corp",
                "sub": "service_node_01",
                "role": "ADMIN",
                "iss": "aaas-node",
                "aud": "aaas-python",
                "exp": int(time.time()) + 3600
            },
            VALID_SERVICE_SECRET,
            algorithm="HS256"
        )
        claims = verify_service_jwt(f"Bearer {token}")
        self.assertEqual(claims["workspace_id"], "ws_acme_corp")
        self.assertEqual(claims["role"], "ADMIN")

    def test_04_service_jwt_rejects_wrong_audience_or_issuer(self):
        """Service JWT rejects invalid audience or issuer with 401."""
        import time
        token = jwt.encode(
            {
                "workspace_id": "ws_acme_corp",
                "sub": "attacker",
                "iss": "wrong-issuer",
                "aud": "wrong-audience",
                "exp": int(time.time()) + 3600
            },
            VALID_SERVICE_SECRET,
            algorithm="HS256"
        )
        with self.assertRaises(HTTPException) as ctx:
            verify_service_jwt(f"Bearer {token}")
        self.assertEqual(ctx.exception.status_code, 401)

    def test_05_order_lookup_requires_matching_customer_email(self):
        """Order lookup requires customer email and matches case-insensitively; mismatch returns not found."""
        # 1. Matching email succeeds
        res = lookup_order("ws_acme_corp", "#10482", "sarah.connor@example.com")
        self.assertTrue(res.get("found"))
        self.assertEqual(res["order"]["order_number"], "#10482")

        # 2. Mismatched email returns not found (no enumeration)
        res_mismatch = lookup_order("ws_acme_corp", "#10482", "attacker@evil.com")
        self.assertFalse(res_mismatch.get("found"))

        # 3. Missing email returns not found
        res_missing = lookup_order("ws_acme_corp", "#10482", "")
        self.assertFalse(res_missing.get("found"))

    def test_06_empty_tenant_rag_returns_zero_citations(self):
        """Empty workspace returns 0 citations and no invented policy text."""
        res = execute_rag_pipeline(
            question="What is the store return and warranty policy?",
            workspace_id="ws_empty_tenant",
            tenant_chunks=[]
        )
        self.assertEqual(len(res.get("citations", [])), 0)
        self.assertIn("do not have", res["natural_answer"].lower())

if __name__ == "__main__":
    unittest.main(verbosity=2)
